Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-162.ca-central-1.compute.internal
Sha: f49f1bec671f2c7af2a556dd865fa580becb1bab
Number of Error(s): 9
/home/centos/smoketest/PR-13625/HPCC-Platform/roxie/ccd/ccdactivities.cpp:1361:5: error: expected \xe2\x80\x98;\xe2\x80\x99 at end of member declaration
/home/centos/smoketest/PR-13625/HPCC-Platform/roxie/ccd/ccdactivities.cpp:1361:23: error: ISO C++ forbids declaration of \xe2\x80\x98helper\xe2\x80\x99 with no type [-fpermissive]
/home/centos/smoketest/PR-13625/HPCC-Platform/roxie/ccd/ccdactivities.cpp:1372:25: error: cannot convert \xe2\x80\x98IHThorCsvReadArg*\xe2\x80\x99 to \xe2\x80\x98int*\xe2\x80\x99 in assignment
/home/centos/smoketest/PR-13625/HPCC-Platform/roxie/ccd/ccdactivities.cpp:1374:17: error: request for member \xe2\x80\x98setCallback\xe2\x80\x99 in \xe2\x80\x98*((CsvRecordProcessor*)this)->CsvRecordProcessor::helper\xe2\x80\x99, which is of non-class type \xe2\x80\x98int\xe2\x80\x99
/home/centos/smoketest/PR-13625/HPCC-Platform/roxie/ccd/ccdactivities.cpp:1380:44: error: request for member \xe2\x80\x98queryCsvParameters\xe2\x80\x99 in \xe2\x80\x98*((CsvRecordProcessor*)this)->CsvRecordProcessor::helper\xe2\x80\x99, which is of non-class type \xe2\x80\x98int\xe2\x80\x99
/home/centos/smoketest/PR-13625/HPCC-Platform/roxie/ccd/ccdactivities.cpp:1398:34: error: request for member \xe2\x80\x98getMaxColumns\xe2\x80\x99 in \xe2\x80\x98*((CsvRecordProcessor*)this)->CsvRecordProcessor::helper\xe2\x80\x99, which is of non-class type \xe2\x80\x98int\xe2\x80\x99
make[2]: *** [roxie/ccd/CMakeFiles/ccd.dir/build.make:63: roxie/ccd/CMakeFiles/ccd.dir/ccdactivities.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:16058: roxie/ccd/CMakeFiles/ccd.dir/all] Error 2
make: *** [Makefile:163: all] Error 2

[Build error:](https://gist.github.com/4eb7fdf16b955f7559e0e6b5f82088e5)
Build log [Build log](https://gist.github.com/HPCCSmoketest/4eb7fdf16b955f7559e0e6b5f82088e5#file-result-20-04-14-14-09-06-log)

--- a/roxie/ccd/ccdactivities.cpp
+++ b/roxie/ccd/ccdactivities.cpp
@@ -1098,6 +1098,7 @@ public:
 
 protected:
     IHThorCsvReadArg *helper;
+    Owned<IDirectReader> ownedReader; // Ensure that the byte stream reader is released
     const IResolvedFile *datafile;
     size32_t maxRowSize;
 
@@ -1119,9 +1120,12 @@ public:
     {
         {
             CriticalBlock p(pcrit);
+            //Processor should be clear when this is called, but include inside the critical block to be safe
+            processor.clear();
+            ownedReader.setown(manager->createReader(postFilter, false, readPos, parallelPartNo, numParallel, translators));
             processor.setown(
                     createCsvRecordProcessor(*this,
-                                             manager->createReader(postFilter, false, readPos, parallelPartNo, numParallel, translators),
+                                             ownedReader,
                                              packet->queryHeader().channel==1 && !resent,
                                              varFileInfo ? varFileInfo.get() : datafile, maxRowSize));
         }
@@ -1145,6 +1149,7 @@ public:
 
 protected:
     IHThorXmlReadArg *helper;
+    Owned<IDirectReader> ownedReader; // Ensure that the byte stream reader is released
 
 public:
     CRoxieXmlReadActivity(SlaveContextLogger &_logctx, IRoxieQueryPacket *_packet, HelperFactory *_hFactory, const CSlaveActivityFactory *_aFactory,
@@ -1164,7 +1169,9 @@ public:
     {
         {
             CriticalBlock p(pcrit);
-            processor.setown(createXmlRecordProcessor(*this, manager->createReader(postFilter, false, readPos, parallelPartNo, numParallel, translators)));
+            processor.clear();
+            ownedReader.setown(manager->createReader(postFilter, false, readPos, parallelPartNo, numParallel, translators));
+            processor.setown(createXmlRecordProcessor(*this, ownedReader));
         }
         unsigned __int64 rowLimit = helper->getRowLimit();
         unsigned __int64 stopAfter = helper->getChooseNLimit();

